jupyter-extension/jupyterlab_pachyderm/pps_client.py

```diff
@@ -0,0 +1,79 @@
+import os.path
+import json
+import subprocess
+from tempfile import NamedTemporaryFile
+
+
+from tornado.httpclient import AsyncHTTPClient
+
+from .log import get_logger
+
+
+class PPSClient:
+    """Client interface for the PPS extension backend."""
+
+    def __init__(self):
+        self.client = AsyncHTTPClient()
+
+    async def generate(self, path, config):
+        """Generates the pipeline spec from the Notebook file specified.
+
+        Args:
+            path: The path (within Jupyter) to the Notebook file.
+            config: The PPS configuration for the Notebook file.
+        """
+        get_logger().debug(f"path: {path} | body: {config}")
+        # script = await self.client.fetch(f"http://localhost:8888/nbconvert/python{path}")
+        script_name = os.path.basename(path)
+        pipeline_spec = create_pipeline_spec("test_pipeline", "python:3.10", dict(pfs=dict(repo="data")), script_name)
+        return json.dumps(pipeline_spec)
+
+    async def create(self, path, config):
+        """Creates the pipeline from the Notebook file specified.
+
+        Args:
+            path: The path (within Jupyter) to the Notebook file.
+            config: The PPS configuration for the Notebook file.
+        """
+        get_logger().debug(f"path: {path} | body: {config}")
+        input_spec = config.get("input")
+        with NamedTemporaryFile() as temp_config:
+            # TODO: Write the SAME config as a yaml file
+            subprocess.run(
+                ["same", "run", "--same-file", temp_config.name, "--target", "pachyderm", "--input", str(input_spec)]
```

Comment:
This connects to pachyderm using pachyderm's `config.json` on the host, correct? This will work in combined mode, but we'll need to figure out what to do in sidecar mode.

Reply:
In sidecar mode the configuration for the mount extension is stored in the sidecar. This means that you're technically not logged into pachctl inside the jupyter container. You would need to run `pachctl connect <address>` and potentially `pachctl auth login` to prior to this same command running.

Reply:
Yea, sidecar mode would force another login for the pps extension :/. You can get the cluster address from the mount server.